[zeus] libcxx fails to build for armv5e

I try to generate a SDK that targets the quite old armv5e architecture (`require conf/machine/include/tune-arm926ejs.inc`). libcxx fails for this arch, while it is fine for all other archs I need (armv7, powerpc, x86) with the following error at link:

`
/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-clang++ -fPIC -march=armv5te -marm   -mlittle-endian -Wno-error=unused-command-line-argument -Qunused-arguments  --sysroot=/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0=/usr/src/debug/libcxx/9.0.1-r0                      -fdebug-prefix-map=/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0=/usr/src/debug/libcxx/9.0.1-r0                      -fdebug-prefix-map=/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot=                      -fdebug-prefix-map=/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native=   -fvisibility-inlines-hidden  -march=armv5te -marm   -mlittle-endian -Wno-error=unused-command-line-argument -Qunused-arguments  --sysroot=/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot -mfpu=vfp2 -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers  -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -g  -latomic -Wl,-z,defs -Wl,-z,nodelete  -L/home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/build/lib -shared -Wl,-soname,libc++.so.1 -o lib/libc++.so.1.0 projects/libcxx/src/CMakeFiles/cxx_shared.dir/algorithm.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/any.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/bind.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/charconv.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/chrono.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/condition_variable.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/condition_variable_destructor.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/debug.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/exception.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/functional.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/future.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/hash.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/ios.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/iostream.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/locale.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/memory.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/mutex.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/mutex_destructor.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/new.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/optional.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/random.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/regex.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/shared_mutex.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/stdexcept.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/string.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/strstream.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/system_error.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/thread.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/typeinfo.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/utility.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/valarray.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/variant.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/vector.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/filesystem/operations.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/filesystem/directory_iterator.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib" -nodefaultlibs -lpthread -lc -lm -lrt /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot/usr/lib/clang/9.0.1/lib/linux/libclang_rt.builtins-arm.a -lc++abi && :
| /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-ld: projects/libcxx/src/CMakeFiles/cxx_shared.dir/locale.cpp.o: in function `std::__1::__cloc()':
| /home/chvalon/wcom-build-dir/tmp-glibc/work-shared/llvm-project-source-9.0.1-r0/git/libcxx/src/locale.cpp:72: undefined reference to `__sync_val_compare_and_swap_1'
| /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-ld: /home/chvalon/wcom-build-dir/tmp-glibc/work-shared/llvm-project-source-9.0.1-r0/git/libcxx/src/locale.cpp:72: undefined reference to `__sync_val_compare_and_swap_1'
| /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-ld: /home/chvalon/wcom-build-dir/tmp-glibc/work-shared/llvm-project-source-9.0.1-r0/git/libcxx/src/locale.cpp:72: undefined reference to `__sync_val_compare_and_swap_1'
| /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-ld: /home/chvalon/wcom-build-dir/tmp-glibc/work-shared/llvm-project-source-9.0.1-r0/git/libcxx/src/locale.cpp:72: undefined reference to `__sync_val_compare_and_swap_1'
| /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-ld: /home/chvalon/wcom-build-dir/tmp-glibc/work-shared/llvm-project-source-9.0.1-r0/git/libcxx/src/locale.cpp:72: undefined reference to `__sync_val_compare_and_swap_1'
| /home/chvalon/wcom-build-dir/tmp-glibc/work/armv5e-oe-linux-gnueabi/libcxx/9.0.1-r0/recipe-sysroot-native/usr/bin/arm-oe-linux-gnueabi/arm-oe-linux-gnueabi-ld: projects/libcxx/src/CMakeFiles/cxx_shared.dir/locale.cpp.o:/home/chvalon/wcom-build-dir/tmp-glibc/work-shared/llvm-project-source-9.0.1-r0/git/libcxx/src/locale.cpp:491: more undefined references to `__sync_val_compare_and_swap_1' follow
| clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
| ninja: build stopped: subcommand failed.
| WARNING: exit code 1 from a shell command.
|
ERROR: Task (/home/chvalon/yocto-tree/poky/meta-clang/recipes-devtools/clang/libcxx_git.bb:do_compile) failed with exit code '1'

`

It's very easy to reproduce: `MACHINE=qemuarmv5 bitbake libcxx`

I have noticed that libcxx sets `PACKAGECONFIG[no-atomics]` for armv5, I have the impression that this is a problem in the same area.

I'm currently working with zeus, so it's version 9, with master it works.

--- a/conf/nonclangable.conf
+++ b/conf/nonclangable.conf
@@ -217,6 +217,8 @@ CPPFLAGS_append_pn-memcached_toolchain-clang = " -Wno-error=embedded-directive"
 #| clang-7: error: assembler command failed with exit code 1 (use -v to see invocation)
 TUNE_CCARGS_remove_pn-upm_toolchain-clang = "-no-integrated-as"
 TUNE_CCARGS_remove_pn-omxplayer_toolchain-clang = "-no-integrated-as"
+TUNE_CCARGS_remove_pn-nfs-utils_toolchain-clang = "-Wno-error=unused-command-line-argument -Qunused-arguments"
+TUNE_CCARGS_append_pn-nfs-utils_toolchain-clang = " -Werror=unknown-warning-option"
 
 #| /usr/src/debug/ruby/2.5.1-r0/build/../ruby-2.5.1/process.c:7073: undefined reference to `__mulodi4'
 #| clang-7: error: linker command failed with exit code 1 (use -v to see invocation)

--- a/recipes-devtools/clang/clang/0028-Add-libgcc-to-link-step-for-libcxx.patch
+++ b/recipes-devtools/clang/clang/0028-Add-libgcc-to-link-step-for-libcxx.patch
@@ -1,0 +1,30 @@
+From 397bd558b83ce7ff3ea69c1c8fb6f36c624b9ca6 Mon Sep 17 00:00:00 2001
+From: Jeremy Puhlman <jpuhlman@mvista.com>
+Date: Thu, 16 Jan 2020 21:16:10 +0000
+Subject: [PATCH] Add libgcc to link step for libcxx
+
+This corrects "undefined reference to __divti3"
+
+Upstream-Status: Inappropriate [configuration]
+
+Signed-off-by: Jeremy Puhlman <jpuhlman@mvista.com>
+---
+ libcxx/src/CMakeLists.txt | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/libcxx/src/CMakeLists.txt b/libcxx/src/CMakeLists.txt
+index 31cd24333a5..d8ae826d7f5 100644
+--- a/libcxx/src/CMakeLists.txt
++++ b/libcxx/src/CMakeLists.txt
+@@ -234,7 +234,7 @@ if (LIBCXX_ENABLE_SHARED)
+     llvm_setup_rpath(cxx_shared)
+   endif()
+   cxx_link_system_libraries(cxx_shared)
+-  target_link_libraries(cxx_shared PRIVATE ${LIBCXX_LIBRARIES})
++  target_link_libraries(cxx_shared PRIVATE ${LIBCXX_LIBRARIES} "$$($$CC --print-libgcc-file-name)")
+   set_target_properties(cxx_shared
+     PROPERTIES
+       COMPILE_FLAGS "${LIBCXX_COMPILE_FLAGS}"
+-- 
+2.13.3
+

--- a/recipes-devtools/clang/clang_git.bb
+++ b/recipes-devtools/clang/clang_git.bb
@@ -229,10 +229,11 @@ FILES_${PN} += "\
   ${datadir}/opt-viewer/ \
 "
 
-FILES_${PN}-libllvm += "\
+FILES_${PN}-libllvm =+ "\
   ${libdir}/libLLVM-${MAJOR_VER}.${MINOR_VER}.so \
   ${libdir}/libLLVM-${MAJOR_VER}.so \
-  ${libdir}/libLLVM-${MAJOR_VER}.${MINOR_VER}svn.so \
+  ${libdir}/libLLVM-${MAJOR_VER}git.so \
+  ${libdir}/libLLVM-${MAJOR_VER}.${MINOR_VER}git.so \
 "
 
 FILES_libclang = "\

--- a/recipes-devtools/clang/common.inc
+++ b/recipes-devtools/clang/common.inc
@@ -35,6 +35,7 @@ SRC_URI = "\
     file://0025-llvm-Let-llvm-ar-name-contain-lib.patch \
     file://0026-libclang-Use-CMAKE_DL_LIBS-for-deducing-libdl.patch \
     file://0027-Fix-sanitizer-common-build-with-glibc-2.31.patch \
+    file://0028-Add-libgcc-to-link-step-for-libcxx.patch \
 "
 
 # Fallback to no-PIE if not set

--- a/recipes-devtools/clang/compiler-rt_git.bb
+++ b/recipes-devtools/clang/compiler-rt_git.bb
@@ -19,7 +19,7 @@ TUNE_CCARGS_remove = "-no-integrated-as"
 
 INHIBIT_DEFAULT_DEPS = "1"
 
-DEPENDS += "ninja-native clang-cross-${TARGET_ARCH} virtual/${MLPREFIX}libc virtual/${TARGET_PREFIX}compilerlibs"
+DEPENDS += "ninja-native clang-cross-${TARGET_ARCH} virtual/${MLPREFIX}libc virtual/${TARGET_PREFIX}compilerlibs virtual/crypt"
 DEPENDS_append_class-nativesdk = " clang-native"
 
 HF = "${@ bb.utils.contains('TUNE_CCARGS_MFLOAT', 'hard', 'hf', '', d)}"

--- a/recipes-devtools/clang/openmp_git.bb
+++ b/recipes-devtools/clang/openmp_git.bb
@@ -21,6 +21,13 @@ inherit cmake pkgconfig perlnative
 
 EXTRA_OECMAKE = "-G Ninja ${S}/openmp"
 
+PACKAGECONFIG ?= "ompt-tools"
+PACKAGECONFIG_remove_arm = "ompt-tools"
+PACKAGECONFIG_remove_mipsarch = "ompt-tools"
+PACKAGECONFIG_remove_powerpc = "ompt-tools"
+
+PACKAGECONFIG[ompt-tools] = "-DOPENMP_ENABLE_OMPT_TOOLS=ON,-DOPENMP_ENABLE_OMPT_TOOLS=OFF,"
+
 do_compile() {
 	ninja ${PARALLEL_MAKE}
 }

